neps/nep-0366.md

```diff
@@ -0,0 +1,62 @@
+---
+NEP: 366
+Title: Meta Transactions
+Author: Illia Polosukhin <ilblackdragon@gmail.com>
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/366
+Status: Draft
+Type: Protocol Track
+Category: Runtime
+Created: 21-Jun-2022
+---
+
+## Summary
+
+In-protocol meta transactions allowing for third-party account to initiate and pay transaction fees on behalf of the account.
+
+## Motivation
+
+NEAR has been designed with simplify of onboarding in mind. One of the large hurdles right now is that after creating an implicit or even named account user doesn't have NEAR to pay gas fees to interact with apps.
+
+For example, apps that pay user for doing work (like NEARCrowd or Sweatcoin) or free-to-play games.
+
+[Aurora Plus](https://aurora.plus) has shown viability of the relayers that can offer some number of free transactions and a subscription model. Shifting the complexity of dealing with fees to the infrastructure from the user space.
+
+## Rationale and alternatives
+
+Proposed here design provides the easiest for users and developers way to onboard and pay for user transactions.
+There is no requirement on the user account, including user account may not even exist on chain and implicit account can be used.
+
+An alterantive is a proxy contracts deployed on the user account. 
+This design has severe limitations as it requires user to deploy such contract and additional costs for storage.
+
+## Specification
+
+Next changes to protocol are required:
+ - New Action kind: `DelegateAction(receiver_id, Action, signer_pk, signature)`
+ - On conversion to Receipt, such action is sent as `DelegateAction` to the `receiver_id`.
+ - On processing Receipt, it verifies the `signature` over the given account and `signer_pk`. Unwraps into `receiver_id: AccountId`, `action: Action` inside into new Receipt with given `receiver_id` and `action`. This means that `predeccessor_id` of such action has been overriden.
+
```

Comment:
This part needs to be expanded on how it works in details. Today when we convert a transaction to a receipt, we directly clone the actions from the transaction https://github.com/near/nearcore/blob/b6b9135e847b9e0070af77bde8caed62ee4c7e57/runtime/runtime/src/lib.rs#L217 and that won't work here. Meta transaction needs to have a separate logic that instead of clone what the `actions` field inside a transaction, it unwraps the transaction and construct a new receipt with actions **inside** the `DelegateAction` action. This means likely we should not allow `DelegateAction` to be mixed with other actions in the same transaction. It almost feels to me that we should make `actions` an enum
```rust
pub enum Actions {
     // regular transaction
     RegularActions(Vec<Action>),
     // meta transaction
     DelegateAction { ...}
}
```
to make it easier to process at runtime

Reply:
The flow is next:
 - User signs `DelegatedActionMessage` and sends to relayer
 - Relayer send `Transaction { ... actions: [DelegatedAction { ... }] }`
 - Transaction gets converted into Receipt with a copy of `DelegatedAction { ... }` out of actions. Note that other actions can be before or after this one, for example if the receiver account has a contract and some function needs to be called  first, that is also possible.
 - The Receipt with `DelegatedAction` arrives to the recipient account and gets converted into regular actions per described rules in the spec.

Hence the direct cloning of actions from Transaction works as expected. The conversion happens on the processing of `DelegateAction` state already on the recipient account.

I'll add this to the proposal so it's clearer.